BUG: Handle item is None in certain conditions

pllim opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request is to not let Imviz crash in certain use cases. I am not sure how to write unit test for this.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

Fixes #689 (please double check, I cannot consistently get this error myself)
Fixes #1614 

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [ ] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [ ] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

kecnry commented on jdaviz/configs/default/plugins/viewers.py:
```diff
@@ -124,6 +124,8 @@ def _on_layers_update(self, layers=None):
             return
 
         viewer_item = self.jdaviz_app._viewer_item_by_id(self.reference_id)
+        if viewer_item is None:
+            return
```
this gets triggered when the viewer is removed from the app before the layer update messages are processed?  If so, I think this makes sense as a good way to quickly prevent that case from causing issues.  We don't need to worry about any of the following logic (which only affects things in the viewer, not higher in the stack) since the viewer is no longer in the app.

pllim replied:
This one is for #1614 . I don't grok all the callback stuff but when this happens:

```
  114 viewer_item = self.jdaviz_app._viewer_item_by_id(self.reference_id)
 --> 115 selected_data_items = viewer_item['selected_data_items']
```

`self.reference_id` actually points to the deleted viewer ID (not sure why, maybe upstream bug). So I think returning here is good enough until we find a better fix.

kecnry commented on jdaviz/app.py:
```diff
@@ -1299,7 +1299,11 @@ def _update_selected_data_items(self, viewer_id, selected_items):
         for data_id, visibility in selected_items.items():
-            label = self._get_data_item_by_id(data_id)['name']
+            data_item = self._get_data_item_by_id(data_id)
+            if data_item is None:
+                label = None
```
what scenario triggers this case?  I'm just trying to think if there's a way we can prevent it entirely, or if not if the warning that it will raise is warranted or will be to noisy.

pllim replied:
For this one, I cannot reproduce it, so I just looked at the traceback at https://github.com/spacetelescope/jdaviz/issues/689#issuecomment-1245931466 and did my best guess.

kecnry replied:
ok, at least its not a common scenario then.  I don't think the extra logic to prevent an error is a problem... as long as it doesn't result in a lot of warnings for the user.  But I guess we'll at least have to reproduce (or find someone who can) before closing that issue.  I'll try to do some testing before approving, but otherwise the code changes themselves look good to me!  Thanks!

kecnry reviewed (approved):
I was able to reproduce #689 on main.  This PR does prevent the traceback, but I still don't think the workflow in #689 is handled correctly.  My vote would be to merge this PR as-is but NOT close that issue (which I'll update with what I'm seeing and how to reproduce) with this PR.

(Regression test coverage would be nice, but I'm not sure if API calls to remove viewers would have the signals processed in the same order to reproduce the traceback or not... it might be worth a shot though!)

pllim commented:
@kecnry , you mean regression test for #1614?

kecnry commented:
I suppose both really (coverage for the new lines of code to ensure they avoid those tracebacks)

pllim commented:
@kecnry , for #689, I don't know how to consistently reproduce it but looks like you did? If so, please give me the workflow and I'll see if I can turn that into a regression test.

pllim commented:
@kecnry , so I think only 2 remaining things:

1. I need a workflow if you want me to add a test for #689. I already added one for #1614 here.
2. I also added logic to handle https://github.com/spacetelescope/jdaviz/issues/689#issuecomment-1249665606 . With this addition, would this PR now actually close that issue or you still want to keep it open?

kecnry commented:
@pllim - is the workflow I added in #689 not sufficient for the test?  It looks like your change to appropriately handle the index does fix the original problem - I wonder if the `if data_item is None` check should now be removed so that if any other bug uncovers a similar situation it won't silently be ignored and difficult to track down?  Or do you think its better to leave that safeguard in to avoid tracebacks for the user?

pllim commented:
> if the if `data_item is None` check should now be removed

Hmm... If it is not needed for *this* fix, I guess we can remove it. Lemme see.

pllim commented:
> is the workflow I added in https://github.com/spacetelescope/jdaviz/issues/689 not sufficient for the test?

I think there are 2 different problems being discussed at Issue 689.

* I fixed the second duplicate label being overwritten and that is just label generation and should be already tested in this PR via `test_helpers.py` changes.
* I never was able to reproduce the traceback reported at https://github.com/spacetelescope/jdaviz/issues/689#issuecomment-1245931466

rosteen reviewed (approved):
Looks good to me, I confirmed that both of the linked bugs are fixed by this. Thanks!

pllim commented:
Thanks for the reviews!
